File: app/src/main/java/jp/hazuki/yuzubrowser/action/manager/SoftButtonActionArrayFile.java
package jp.hazuki.yuzubrowser.action.manager;

import android.content.Context;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import jp.hazuki.yuzubrowser.action.ActionFile;
import jp.hazuki.yuzubrowser.action.SingleAction;

public class SoftButtonActionArrayFile extends ActionFile {
    private static final long serialVersionUID = -8451972340596132660L;

    private final String FOLDER_NAME;
    private final int id;
    public final List<SoftButtonActionFile> list = new ArrayList<>();

    public SoftButtonActionArrayFile(String folder_name, int id) {
        this.id = id;
        FOLDER_NAME = folder_name;
    }

    public void add(SingleAction action) {
        SoftButtonActionFile array = new SoftButtonActionFile();
        array.press.add(action);
        list.add(array);
    }

    public void add(SoftButtonActionFile action) {
        SoftButtonActionFile array = new SoftButtonActionFile();
        array.press.addAll(action.press);
        array.lpress.addAll(action.lpress);
        array.up.addAll(action.up);
        array.down.addAll(action.down);
        array.left.addAll(action.left);
        array.right.addAll(action.right);
        list.add(array);
    }

    public SoftButtonActionFile getActionList(int no) {
        if (no < 0)
            throw new IllegalArgumentException("no < 0");
        expand(no + 1);
        return list.get(no);
    }

    public int expand(int size) {
        if (size < 0)
            throw new IllegalArgumentException("size < 0");
        for (int i = list.size() - size; i < 0; ++i)
            list.add(new SoftButtonActionFile());
        return list.size();
    }

    @Override
    public File getFile(Context context) {
        return new File(context.getDir(FOLDER_NAME, Context.MODE_PRIVATE), id + ".dat");
    }

    @Override
    public void reset() {
        list.clear();
    }

    @Override
    public boolean load(JsonParser parser) throws IOException {
        if (parser.nextToken() != JsonToken.START_ARRAY) return false;
        for (; ; ) {
            SoftButtonActionFile action = new SoftButtonActionFile();
            if (!action.load(parser)) {
                if (parser.getCurrentToken() == JsonToken.END_ARRAY)
                    break;
                else
                    return false;
            }
            list.add(action);
        }
        return true;
    }

    @Override
    public boolean write(JsonGenerator generator) throws IOException {
        generator.writeStartArray();
        for (SoftButtonActionFile action : list) {
            action.write(generator);
        }
        generator.writeEndArray();
        return true;
    }

}

File: legacy/src/main/java/jp/hazuki/yuzubrowser/legacy/action/manager/SoftButtonActionArrayFile.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.legacy.action.manager

import android.content.Context
import com.squareup.moshi.JsonReader
import com.squareup.moshi.JsonWriter
import jp.hazuki.yuzubrowser.legacy.action.ActionFile
import jp.hazuki.yuzubrowser.legacy.action.SingleAction
import java.io.File
import java.io.IOException
import java.util.*

class SoftButtonActionArrayFile(private val FOLDER_NAME: String, private val id: Int) : ActionFile() {
    val list: MutableList<SoftButtonActionFile> = ArrayList()

    fun add(action: SingleAction) {
        val array = SoftButtonActionFile()
        array.press.add(action)
        list.add(array)
    }

    fun add(action: SingleAction, longPress: SingleAction) {
        val array = SoftButtonActionFile()
        array.press.add(action)
        array.lpress.add(longPress)
        list.add(array)
    }

    fun add(action: SoftButtonActionFile) {
        val array = SoftButtonActionFile()
        array.press.addAll(action.press)
        array.lpress.addAll(action.lpress)
        array.up.addAll(action.up)
        array.down.addAll(action.down)
        array.left.addAll(action.left)
        array.right.addAll(action.right)
        list.add(array)
    }

    fun getActionList(no: Int): SoftButtonActionFile {
        if (no < 0)
            throw IllegalArgumentException("no < 0")
        expand(no + 1)
        return list[no]
    }

    fun expand(size: Int): Int {
        if (size < 0)
            throw IllegalArgumentException("size < 0")
        for (i in list.size - size..-1)
            list.add(SoftButtonActionFile())
        return list.size
    }

    override fun getFile(context: Context): File {
        return File(context.getDir(FOLDER_NAME, Context.MODE_PRIVATE), id.toString() + ".dat")
    }

    override fun reset() {
        list.clear()
    }

    @Throws(IOException::class)
    override fun load(reader: JsonReader): Boolean {
        if (reader.peek() != JsonReader.Token.BEGIN_ARRAY) return false
        reader.beginArray()
        while (reader.hasNext()) {
            val action = SoftButtonActionFile()
            if (!action.load(reader)) {
                return if (reader.peek() == JsonReader.Token.END_ARRAY)
                    break
                else
                    false
            }
            list.add(action)
        }
        reader.endArray()
        return true
    }

    @Throws(IOException::class)
    override fun write(writer: JsonWriter): Boolean {
        writer.beginArray()
        list.forEach { it.write(writer) }
        writer.endArray()
        return true
    }

    companion object {
        private const val serialVersionUID = -8451972340596132660L
    }

}
